### PR TITLE
Connecteur Workday (sans infos sensibles)

### PR DESCRIPTION
**Pull request by pcadoret**
Incluant Query Parameters dynamiques (ne pas passer les paramètres si non définis)

### CODE REVIEW

File: tests/workday/test_workday.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        service_WSDL_URL='https://wd3-impl-services1.workday.com/ccx/service/umanis/Human_Resources/v33.2',
+        operation='Get_Workers',
+        request_references_param={'WrongQuery' : 'WrongParam'},
+        filter="[.Worker[].Worker_Data | {User_ID: .User_ID, Worker_ID: .Worker_ID}]"
+    )
+
+    with pytest.raises(TypeError):
+        connector.get_df(data_sourceWrong)
+
+""" This test is not pertinent, as the "Service" name is an arbitrary string, just a key to store the WSDL URL in a dict
```
Comment (piotch) [inline tests/workday/test_workday.py]: use `@pytest.mark.skip(reason="...")`

File: toucan_connectors/workday/workday_connector.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    if isinstance(obj, (datetime, date)):
+        return obj.isoformat()
+    if isinstance(obj, decimal.Decimal):
+        return (str(obj) for obj in [obj])
+    if isinstance(obj, types.GeneratorType):
+        return list(obj)
+    raise TypeError ("Type %s not serializable" % type(obj))
+
+def use_jq_to_parse(data, jq_filter):
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py]: not sure about the need to wrap `transform_with_jq` also worried about logging in `error` level raw customers data, I would rather not do this

File: toucan_connectors/workday/workday_connector.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        description='Check "Public Web Services" page in Workday',
+        example='Human_Resources'
+    )
+    service_WSDL_URL: str = Field(
+        None,
+        title='URL of the WSDL of the Workday API Service you want to use.',
+        description='https://{Workday domain}/ccx/service/{tenant}/{service}/{version}?wsdl',
+        example='https://wd3-impl-services1.workday.com/ccx/service/umanis/Human_Resources/v33.2?wsdl'
+    )
+    operation: str = Field(
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py]: enum possible here ?

File: toucan_connectors/workday/workday_connector.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+		apis = {
+			data_source.service: data_source.service_WSDL_URL
+		}
+
+		client = workday.WorkdayClient(
+			wsdls=apis,
+			authentication=auth
+		)
+		#print(client)
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py]: remove debug prints

File: toucan_connectors/workday/workday_connector.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+		query_arguments = dict()
+		if not (data_source.request_references_param is None) and len(data_source.request_references_param) > 0:
+			query_arguments.update(Request_References = data_source.request_references_param)
+		if not (data_source.request_criteria_param is None) and len(data_source.request_criteria_param) > 0:
+			query_arguments.update(Request_Criteria = data_source.request_criteria_param)
+		if not (data_source.response_filter_param is None) and len(data_source.response_filter_param) > 0:
+			query_arguments.update(Response_Filter = data_source.response_filter_param)
+		if not (data_source.response_group_param is None) and len(data_source.response_group_param) > 0:
+			query_arguments.update(Response_Group = data_source.response_group_param)
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py]: I don't understand this code, could you tell me why you are doing this ? is there not a simpler way to do it ? if not could you comment ?

File: toucan_connectors/workday/workday_connector.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+		data = getattr(getattr(client, data_source.service) , data_source.operation)(query_arguments)
+
+		num_page = data.total_pages
+
+		data_json = json.loads(json.dumps(zeep.helpers.serialize_object(data.data), default=json_serial))
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py]: could you explain why there is a need to `serialize_object` -> `dumps` -> `loads` that seems very costly

Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py:143]: zeep.helpers.serialize_object() returns a dict [https://docs.python-zeep.org/en/master/helpers.html](https://docs.python-zeep.org/en/master/helpers.html), so we have to convert this dict to a json. Ok for you ?

File: toucan_connectors/workday/workday_connector.py (hunk 2 above)
Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py:30]: we are deleting this wrapping function

File: toucan_connectors/workday/workday_connector.py (hunk 3 above)
Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py:42]: No there is too much possibilities

File: toucan_connectors/workday/workday_connector.py (hunk 5 above)
Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py]: We deleted this part

File: toucan_connectors/workday/workday_connector.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        client = workday.WorkdayClient(wsdls=apis, authentication=auth)
+
+        data = getattr(getattr(client, data_source.service), data_source.operation)(
+            **data_source.request_parameters
+        )
+
+        num_page = data.total_pages
+
+        data_json = json.loads(
+            json.dumps(zeep.helpers.serialize_object(data.data), default=json_serial)
```
Comment (piotch) [inline toucan_connectors/workday/workday_connector.py:100]: I still don't get it : 

```
>>> import json
>>> input = {}
>>> json.loads(json.dumps(input)) == input
True
```

File: setup.py [hunk 8]
```diff
@@ -31,6 +31,8 @@
     'sap_hana': ['pyhdb>=0.3.4'],
     'snowflake': ['snowflake-connector-python'],
     'toucan_toco': ['toucan_client'],
+    'workday': zeep,
+    'workday': workday
```
Comment (piotch) [inline setup.py:35]:
```suggestion
    'workday': ['zeep', 'workday],
```

File: toucan_connectors/workday/workday_connector.py (hunk 7 above)
Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py:100]: In our case, we do are using a custom JSONEncoder that overrides the default() method to serialize additional types. The zeep.helpers.serialize_object method doesn't return a serializable object.

Ok for you ?

Comment (piotch) [reply toucan_connectors/workday/workday_connector.py:100]: You are returning a DataFrame and not JSON text so why is it a problem to have dates and decimal ?

Comment (CharlesRngrd) [reply toucan_connectors/workday/workday_connector.py:100]: Before returning a DataFrame but I need to transform a JSON with jq.

Here is the OrderedDict returned by zeep.helpers.serialize_object() :

> OrderedDict([('Absence_Input', [OrderedDict([('Absence_Input_Reference', OrderedDict([('ID', [OrderedDict([('_value
> _1', '08d4a6121b760154a3c7d6e1600c6646'), ('type', 'WID')])]), ('Descriptor', None)])), ('Absence_Input_Data', Orde
> redDict([('Absence_Input_ID', None), ('Batch_ID', 'Chargement Pris CP1 Golive Umanis'), ('Worker_Reference', Ordere
> dDict([('ID', [OrderedDict([('_value_1', '9836946ba6f4017d017f88c78d212b39'), ('type', 'WID')]), OrderedDict([('_va
> lue_1', 'E5042005'), ('type', 'Employee_ID')])]), ('Descriptor', None)])), ('Position_Reference', None), ('Absence_
> Component_Reference', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a6121b7601ba5f1109f95f0cfa2e'), ('type', 
> 'WID')]), OrderedDict([('_value_1', "FRA - Congés payés en cours d'acquisition (CP1) Time Off Reprise"), ('type', '
> Time_Off_Code')])]), ('Descriptor', None)])), ('Start_Date', datetime.date(2019, 1, 1)), ('End_Date', datetime.date
> (2019, 1, 31)), ('Reference_Date', datetime.date(2019, 1, 1)), ('Hours', Decimal('0')), ('Adjustment', True), ('Com
> ment', None)]))]), OrderedDict([('Absence_Input_Reference', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a61
> 21b76019c685b242a610cea54'), ('type', 'WID')])]), ('Descriptor', None)])), ('Absence_Input_Data', OrderedDict([('Ab
> sence_Input_ID', None), ('Batch_ID', 'Chargement Pris CP2 Golive Umanis'), ('Worker_Reference', OrderedDict([('ID',
>  [OrderedDict([('_value_1', '9836946ba6f401d1dbd602168e2192d4'), ('type', 'WID')]), OrderedDict([('_value_1', 'E508
> 0032'), ('type', 'Employee_ID')])]), ('Descriptor', None)])), ('Position_Reference', None), ('Absence_Component_Ref
> erence', OrderedDict([('ID', [OrderedDict([('_value_1', '08d4a6121b760197da744601600c1d2f'), ('type', 'WID')]), Ord
> eredDict([('_value_1', 'FRA - Congés Payés Acquis (CP2) Time Off Reprise'), ('type', 'Time_Off_Code')])]), ('Descri
> ptor', None)])), ('Start_Date', datetime.date(2019, 1, 1)), ('End_Date', datetime.date(2019, 1, 31)), ('Reference_D
> ate', datetime.date(2019, 1, 1)), ('Hours', Decimal('5')), ('Adjustment', True), ('Comment', None)]))])])])

And here is the error :

> ../../toucan_connectors/toucan_connector.py:162: in get_func_and_retry
>     return self.retry_decorator(func)(self, *args, **kwargs)
> ../../toucan_connectors/toucan_connector.py:236: in get_df
>     res = self._retrieve_data(data_source)
> ../../toucan_connectors/workday/workday_connector.py:142: in _retrieve_data
>     df = pd.DataFrame(use_jq_to_parse(zeep.helpers.serialize_object(data.data), data_source.filter))
> ../../toucan_connectors/workday/workday_connector.py:32: in use_jq_to_parse
>     return transform_with_jq(data, jq_filter)
> ../../toucan_connectors/common.py:273: in transform_with_jq
>     data = jq(jq_filter).transform(data, multiple_output=True)
> jq.pyx:178: in jq._Program.transform
>     ???
> jq.pyx:165: in jq._Program.input
>     ???
> /usr/lib/python3.6/json/__init__.py:231: in dumps
>     return _default_encoder.encode(obj)
> /usr/lib/python3.6/json/encoder.py:199: in encode
>     chunks = self.iterencode(o, _one_shot=True)
> /usr/lib/python3.6/json/encoder.py:257: in iterencode
>     return _iterencode(o, 0)
> 
> TypeError: Object of type 'Decimal' is not JSON serializable

Comment (piotch) [reply toucan_connectors/workday/workday_connector.py:100]: I understand the problem, my point is can we not have a simple / less expensive way to get the type we want : I see for example that the `zeep.helpers.serialize_object` takes an extra `target_cls` argument. cf. https://docs.python-zeep.org/en/master/helpers.html#zeep.helpers.serialize_object have you looked into this ?

**Pull request closed without merging**